src/install/warmViews.js

```diff
@@ -0,0 +1,165 @@
+    indexers.forEach(calculateAverageProgress);
+
+    entry.indexers = indexers;
+
+    logIndexersProgress(indexers);
+    return utils.update(deployDoc);
+  };
+
+  const setTasksToComplete = (indexer) => {
+    Object
+      .keys(indexer.tasks)
+      .forEach(pid => {
+        indexer.tasks[pid] = 100;
+      });
+  };
+
+  const calculateAverageProgress = (indexer) => {
+    const tasks = Object.keys(indexer.tasks);
+    indexer.progress = Math.round(tasks.reduce((progress, pid) => progress + indexer.tasks[pid], 0) / tasks.length);
+  };
+
+  const updateRunningTasks = (indexers, activeTasks = []) => {
+    activeTasks.forEach(task => {
+      let indexer = indexers.find(indexer => indexer.design_document === task.design_document);
+      if (!indexer) {
+        indexer = {
+          design_document: task.design_document,
+          tasks: {},
+        };
+        indexers.push(indexer);
+      }
+      indexer.tasks[`${task.node}-${task.pid}`] = task.progress;
+    });
+  };
+
+  let stopViewWarming = false;
+
+  // Query _active_tasks every ACTIVE_TASK_QUERY_INTERVAL seconds until `stopViewWarming` is true
+  const writeProgressTimeout = (rej) => {
+    setTimeout(() => {
+      if (stopViewWarming) {
+        return;
+      }
+      writeProgress()
+        .then(writeProgressTimeout)
+        .catch(rej);
+    }, ACTIVE_TASK_QUERY_INTERVAL);
+  };
+
+  const probeViews = viewlist => {
+    if (stopViewWarming) {
+      return;
+    }
+
+    return Promise
+      .all(viewlist.map(view => DB.app.query(view, { limit: 1 })))
+      .then(() => {
+        stopViewWarming = true;
+        info('Warming views complete');
+        return updateIndexers();
+      })
+      .catch(err => {
+        if (err.error !== 'timeout') {
+          // Ignore errors in the view warming loop because long-running view queries aren't that
+          // trust-worthy. We *do* check for errors in the writeProgressTimeout loop, so that will
+          // catch real CouchDB errors
+          info(`Unexpected error while warming: (${err.message}), continuing`);
+        }
+
+        return probeViews(viewlist);
+      });
+  };
+
+  const firstView = ddoc =>
+    `${ddoc._id.replace('_design/', '')}/${Object.keys(ddoc.views).find(k => k !== 'lib')}`;
+
+  return utils.getStagedDdocs(true)
+    .then(ddocs => {
+      debug(`Got ${ddocs.length} staged ddocs`);
+      const queries = ddocs
+        .filter(ddoc => ddoc.views && Object.keys(ddoc.views).length)
+        .map(firstView);
+
+      info('Beginning view warming');
+
+      deployDoc.log.push({
+        type: 'warm_log'
+      });
+
+      return utils.update(deployDoc)
+        .then(() => {
+          return new Promise((res, rej) => {
+            writeProgressTimeout(rej);
+            probeViews(queries)
+              .then(() => undefined)
+              .catch(err => err)
+              .then(err => {
+                // Manually implementing something similar to a `finally` block
```

Comment:
This screams of `Promise.race` to me. 
Something like this maybe? Not saying you should change it, but it was a nice exercise for me to do a pseudo implementation.

```
// instead of writeProgressTimeout
const writeProgressPromise = () => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      if (stopViewWarming) {
        return;
      }

      writeProgress().then(writeProgressPromise).then(resolve).catch(reject);
    }, ACTIVE_TASK_QUERY_INTERVAL);
  });
};
.....
Promise
  .race([writeProgressPromise(), probeViews(queries)])
  .catch(...);
```

Reply:
Oh my, these recursive-ish promise chains leak memory so badly. Ah well, a girl can dream...